source/main/gui/panels/GUI_SurveyMap.cpp

```diff
@@ -364,3 +457,41 @@ void SurveyMap::DrawMapIcon(ImVec2 view_pos, ImVec2 view_size, Ogre::Vector2 vie
     ImVec2 text_pos(img_pos.x - (ImGui::CalcTextSize(caption.c_str()).x/2), img_pos.y + 5);
     drawlist->AddText(text_pos, ImGui::GetColorU32(ImGui::GetStyle().Colors[ImGuiCol_Text]), caption.c_str());
 }
+
+ImVec2 SurveyMap::DrawWaypoint(ImVec2 view_pos, ImVec2 view_size, Ogre::Vector2 view_origin,
+                                      std::string const& caption, int i, bool draw_now)
+{
+    Ogre::Vector2 terrn_size_adj = mTerrainSize;
+    if (mMapMode == SurveyMapMode::SMALL)
+    {
+        terrn_size_adj = mTerrainSize * (1.f - mMapZoom);
+    }
+
+    ImVec2 pos1;
+    pos1.x = view_pos.x + ((ai_waypoints[i].x - view_origin.x) / terrn_size_adj.x) * view_size.x;
+    pos1.y = view_pos.y + ((ai_waypoints[i].z - view_origin.y) / terrn_size_adj.y) * view_size.y;
+
+    if (draw_now)
```

Comment:
Create a function `ImVec2 CalcWaypointMapPos(int idx)`, this will get rid of the `draw_now` hack and make this function shorter, you'll just call `drawlist->AddLine(CalcWaypointPos(i), CalcWaypointPos(i+1))` basically.